examples/acoustic/fwi_operators.py

```diff
@@ -93,36 +127,56 @@ def __init__(self, model, damp, data, recin, time_order=2, spc_order=6, **kwargs
 class GradientOperator(Operator):
     def __init__(self, model, damp, data, recin, u, time_order=2, spc_order=6, **kwargs):
         nrec, nt = data.shape
-        dt = model.get_critical_dt()
+        s, h = symbols('s h')
         v = TimeData(name="v", shape=model.get_shape_comp(), time_dim=nt,
-                     time_order=time_order, space_order=spc_order,
+                     time_order=2, space_order=spc_order,
                      save=False, dtype=damp.dtype)
         m = DenseData(name="m", shape=model.get_shape_comp(), dtype=damp.dtype)
         m.data[:] = model.padm()
         v.pad_time = False
-        rec = SourceLike(name="rec", npoint=nrec, nt=nt, dt=dt, h=model.get_spacing(),
-                         coordinates=data.receiver_coords, ndim=len(damp.shape),
-                         dtype=damp.dtype, nbpml=model.nbpml)
-        rec.data[:] = recin
         grad = DenseData(name="grad", shape=m.shape, dtype=m.dtype)
 
         # Derive stencil from symbolic equation
-        eqn = m * v.dt2 - v.laplace - damp * v.dt
-        stencil = solve(eqn, v.backward)[0]
+        if time_order == 2:
+            Lap = v.laplace
+            Lap2 = 0
+            # PDE for information
+            # eqn = m * v.dt2 - Lap - damp * v.dt
+            dt = model.get_critical_dt()
+            gradient_update = Eq(grad, grad - u.dt2 * v.forward)
+        else:
+            Lap = v.laplace
+            Lap2 = 1 / m * v.laplace2
+            Lap2u = - 1 / m**2 * u.laplace2
+            # PDE for information
+            # eqn = m * v.dt2 - Lap - s**2 / 12 * Lap2 + damp * v.dt
+            dt = 1.73 * model.get_critical_dt()
+            gradient_update = Eq(grad, grad -
+                                 (u.dt2 -
+                                  s ** 2 / 12.0 * Lap2u) * v.forward)
```

Comment:
This would indeed clean up everything. But the point of this PR was not a cleanup PR but only to add the 4th order functionallity we have bneen missing for month. I can try to look at it but it would require a major combined rewrite of `Acoustic_codegen` and `fwi_operators` all together.